data_model/src/predicate.rs

```diff
@@ -7,28 +7,69 @@ use core::{fmt::Display, ops::Not};
 use super::*;
 use crate::{IdBox, Name, Value};
 
-/// Struct representing sequence of non-zero length
-#[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, IntoSchema)]
-pub struct NonEmpty<T> {
-    head: Box<T>,
-    tail: Vec<T>,
-}
+mod nontrivial {
+    #![allow(clippy::expect_used)]
+    use super::*;
+    /// Struct representing sequence of length > 2
+    #[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, IntoSchema)]
+    pub struct NonTrivial<T>(Vec<T>);
 
-impl<T> NonEmpty<T> {
-    /// Extend sequence with elements of another non-empty sequence
-    #[inline]
-    fn extend(&mut self, other: Self) {
-        self.tail.push(*other.head);
-        self.tail.extend(other.tail);
+    impl<T> NonTrivial<T> {
+        /// Constructor
+        #[inline]
+        pub fn new(first: T, second: T) -> Self {
+            Self(vec![first, second])
+        }
+
+        /// Extend sequence with elements of another non-empty sequence
+        #[inline]
+        pub fn extend(&mut self, other: Self) {
+            self.0.extend(other.0)
+        }
+
+        /// Append a value to the end of the sequence
+        #[inline]
+        pub fn push(&mut self, value: T) {
+            self.0.push(value)
+        }
+
+        /// Apply a function to every element of sequence
+        #[must_use]
+        #[inline]
+        pub fn transform(self, f: impl FnMut(T) -> T) -> Self {
+            Self(self.0.into_iter().map(f).collect())
+        }
```

Comment:
```suggestion
        pub fn map(self, f: impl FnMut(T) -> T) -> Self {
            Self(self.0.into_iter().map(f).collect())
        }
```

The word `transform` implies a transformation of the container, just like `from` and `into`.